daffodil-io/src/main/scala/org/apache/daffodil/io/InputSource.scala

```diff
@@ -142,8 +145,9 @@ abstract class InputSource {
  *
  * @param inputStream the java.io.Inputstream to read data from
  * @param bucketSize the size of each individual bucket
+ * @param maxNumberOfBuckets the maximum number of buckets that can be created before beginning to remove old ones
  */ 
-class BucketingInputSource(inputStream: java.io.InputStream, bucketSize: Int = 1 << 13)
+class BucketingInputSource(inputStream: java.io.InputStream, bucketSize: Int = 1 << 13, maxNumberOfBuckets: Int = 32768)
```

Comment:
If I'm doing the math right, a bucketsize of 1<<13 is 4096 bytes, that times 32768 comes out to a little over 100MB. So the most we could backtrack by default is 100MB. That doesn't seem totally unreasonble?

A though, maybe instead of maxNumberOfBuckets passed it, we pass in maxCacheSizeInBytes or something and then convert that to maxNumberOfBuckets internally? That way users don't have to do the math I did above.